pkg/river/token/builder/blocks_nesting_test.go

```diff
@@ -0,0 +1,693 @@
+type OutNoDefInPtrBlkOptWithDef struct {
+	Inner *AttrWithDefault `river:"inner,block,optional"`
+}
+
+func (o *OutNoDefInPtrBlkOptWithDef) testCases() []testCase {
+	return []testCase{
+		{
+			name:  "nil",
+			in:    OutNoDefInPtrBlkOptWithDef{},
+			river: "",
+		},
+		{
+			name: "zero value set",
+			in: OutNoDefInPtrBlkOptWithDef{
+				Inner: &AttrWithDefault{},
+			},
+			river: `
+			inner {
+				number = 0
+			}`,
+		},
+		{
+			name: "different value set",
+			in: OutNoDefInPtrBlkOptWithDef{
+				Inner: &AttrWithDefault{Number: otherDefaultNumber},
+			},
+			river: `
+			inner {
+				number = 321
+			}
+		`,
+		},
+		{
+			name: "default value set",
+			in: OutNoDefInPtrBlkOptWithDef{
+				Inner: &AttrWithDefault{Number: defaultNumber},
+			},
+			river: `inner { }`,
+		},
+	}
+}
+
+// OutDiffDefInPtrBlkOptDiffDef - outer with different default, inner struct pointer, optional with a different default value
+type OutDiffDefInPtrBlkOptDiffDef struct {
+	Inner *AttrWithDefault `river:"inner,block,optional"`
+}
+
+func (o *OutDiffDefInPtrBlkOptDiffDef) SetToDefault() {
+	*o = OutDiffDefInPtrBlkOptDiffDef{Inner: &AttrWithDefault{Number: otherDefaultNumber}}
+}
+
+func (o *OutDiffDefInPtrBlkOptDiffDef) testCases() []testCase {
+	return []testCase{
+		//TODO: invariant violated - not clear how to explicitly set the inner block to nil in River
+		//{
+		//	name:  "nil",
+		//	in:    OutDiffDefInPtrBlkOptDiffDef{},
+		//	river: "",
+		//},
+		{
+			name: "zero value set",
+			in: OutDiffDefInPtrBlkOptDiffDef{
+				Inner: &AttrWithDefault{},
+			},
+			river: `
+			inner {
+				number = 0
+			}`,
+		},
+		{
+			name: "different value set",
+			in: OutDiffDefInPtrBlkOptDiffDef{
+				Inner: &AttrWithDefault{Number: 42},
+			},
+			river: `
+			inner {
+				number = 42
+			}
+		`,
+		},
+		{
+			name: "inner default value set",
+			in: OutDiffDefInPtrBlkOptDiffDef{
+				Inner: &AttrWithDefault{Number: defaultNumber},
+			},
+			river: `inner { }`,
+		},
+		{
+			name: "outer default value set",
+			in: OutDiffDefInPtrBlkOptDiffDef{
+				Inner: &AttrWithDefault{Number: otherDefaultNumber},
+			},
+			river: ``,
+		},
+	}
+}
+
+// ========== tests with inner pointer to struct, but not optional ==========
+
+// OutZeroDefInPtrBlkWithDef - outer with zero value default, inner struct pointer, required with a default value
+type OutZeroDefInPtrBlkWithDef struct {
+	Inner *AttrWithDefault `river:"inner,block"`
+}
+
+func (o *OutZeroDefInPtrBlkWithDef) SetToDefault() {
+	*o = OutZeroDefInPtrBlkWithDef{Inner: &AttrWithDefault{}}
+}
+
+func (o *OutZeroDefInPtrBlkWithDef) testCases() []testCase {
+	return []testCase{
+		//{
+		//	//TODO: encoded river doesn't decode correctly due to error: missing required block "inner"
+		//	// seems like we should have failed on encoding to river?
+		//	name: "nil",
+		//	in: OutZeroDefInPtrBlkWithDef{
+		//		Inner: nil,
+		//	},
+		//	river: "",
+		//},
```

Comment:
Another issue: we encode happily to River, but decoding errors with "missing required block". 
Seems like we should error on encoding too?